fix/BannerBase to TS

dhruvv173 opened the pull request:
## Explanation

- Fixes #19181 

## Screenshots/Screencaps

<!-- If you're making a change to the UI, make sure to capture a screenshot or a short video showing off your work! -->

### Before

https://github.com/MetaMask/metamask-extension/assets/79097544/2a2f8168-4875-4632-bff1-4199bcc8729a

### After

https://github.com/MetaMask/metamask-extension/assets/79097544/99c53b48-be00-47af-b24a-7b36b8043394

![image](https://github.com/MetaMask/metamask-extension/assets/79097544/c866b21c-4c3d-409f-a0b4-2e4e2d0abd84)

## Manual Testing Steps

- Run `yarn storybook` on this branch
- Check controls for `BannerBase`
- `yarn jest ui/components/ocmponent-library/banner-base`

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

dhruvv173 committed:
BannerBase to TS

dhruvv173 commented on ui/components/component-library/banner-base/banner-base.tsx:
```diff
@@ -0,0 +1,97 @@
+import { BoxProps, PolymorphicRef } from '../box';
+import { BannerBaseComponent, BannerBaseProps } from './banner-base.types';
+
+export const BannerBase: BannerBaseComponent = React.forwardRef(
+  <C extends React.ElementType = 'div'>(
```
trying to find a fix for this, input would be appreciated
![image](https://github.com/MetaMask/metamask-extension/assets/79097544/199cb66e-89ec-481a-bfe5-795bd065ba8d)

dhruvv173 commented on ui/components/component-library/banner-base/banner-base.types.ts:
```diff
@@ -0,0 +1,65 @@
+  /**
+   * Additional props to pass to the `Text` component used for the `title` text
+   */
+  titleProps?: any;
```
`titleProps?: TextProps<'span'>` might be wrong but this would be better, right?

Ameralameri reviewed (approved):
Approved

dhruvv173 committed:
snapshot updates

dhruvv173 committed:
more snapshot updates

dhruvv173 committed:
addressing type definition error

dhruvv173 committed:
updating eth-sign-modal snapshot

dhruvv173 commented on ui/components/component-library/banner-base/banner-base.types.ts:
```diff
@@ -0,0 +1,65 @@
+  /**
+   * Props for action button (ButtonLink) of the BannerBase below the children
+   */
+  actionButtonProps?: any;
```
`actionButtonProps?: ButtonLinkProps<'button'>;` ?

georgewrmarshall replied:
Hey @dhruvv173, that makes sense to me thanks!

dhruvv173 commented on ui/components/component-library/banner-base/banner-base.types.ts:
```diff
@@ -0,0 +1,65 @@
+  /**
+   * The onClick handler for the action button (ButtonLink)
+   */
+  actionButtonOnClick?: () => void;
```
`(event: React.MouseEvent<HTMLButtonElement>) => void;` for type safety?

georgewrmarshall committed:
Updates to stories, types and adding data-testid

georgewrmarshall committed:
Updating snapshots

georgewrmarshall commented on ui/components/component-library/banner-base/banner-base.types.ts:
```diff
@@ -0,0 +1,81 @@
+ * Makes all props optional so that if a prop object is used not ALL required props need to be passed
+ * TODO: Move to appropriate place in app as this will be highly reusable
+ */
+type MakePropsOptional<T> = {
+  [K in keyof T]?: T[K];
+};
```
This type helper allows for us to use prop objects without needing to pass all required props through the prop object.


https://github.com/MetaMask/metamask-extension/assets/8112138/74ec6ee8-998c-4621-b64b-acb9d00884fd

georgewrmarshall replied:
cc @garrettbear

georgewrmarshall commented on ui/components/component-library/box/box.types.ts:
```diff
@@ -410,6 +410,12 @@ export interface StyleUtilityProps {
+   * TypeScript is complaining about data- attributes which means we need to explicitly define this as a prop.
+   * TODO: Allow data- attributes.
+   */
+  'data-testid'?: string;
```
Temporary fix to allow `data-testid`s being passed to prop objects which is generally what they will be used for

georgewrmarshall reviewed (approved):
LGTM! I made some small updates based on your suggestions while I had your branch checked out. Thanks for your contribution @dhruvv173

georgewrmarshall committed:
Merge branch 'develop' into fix/19181

georgewrmarshall reviewed (changes requested):
Hey @dhruvv173, this PR should fix the linting error https://github.com/MetaMask/metamask-extension/pull/20448 and looks like I missed some snapshots or they were updated in the last merge from develop. I was also thinking that because I added the `childrenWrapperProps` we should write a unit test for that. There should be some examples of this with `titleProps` in the unit tests. Would you mind updating it?

dhruvv173 commented:
> Hey @dhruvv173, this PR should fix the linting error #20448 and looks like I missed some snapshots or they were updated in the last merge from develop. I was also thinking that because I added the `childrenWrapperProps` we should write a unit test for that. There should be some examples of this with `titleProps` in the unit tests. Would you mind updating it?

I have updated the snapshot locally and lint error will be fixed once your PR is merged. About the latter part, I just want to confirm that we want to add a unit test which checks for `childrenWrapperProps` in the component, right? If yes, I will check for examples in unit tests and work on it. Thank you!

dhruvv173 committed:
Merge branch 'develop' into fix/19181

georgewrmarshall commented:
> > Hey @dhruvv173, this PR should fix the linting error #20448 and looks like I missed some snapshots or they were updated in the last merge from develop. I was also thinking that because I added the `childrenWrapperProps` we should write a unit test for that. There should be some examples of this with `titleProps` in the unit tests. Would you mind updating it?
> 
> I have updated the snapshot locally and lint error will be fixed once your PR is merged. About the latter part, I just want to confirm that we want to add a unit test which checks for `childrenWrapperProps` in the component, right? If yes, I will check for examples in unit tests and work on it. Thank you!

Yes, that's correct! I noticed that we aren't actually testing for `titleProps` or `descriptionProps` we can improve the coverage by adjusting each unit test for these props slightly

```
  it('should render bannerbase title', () => {
    const { getByText } = render(
      <BannerBase
        title="Bannerbase title test"
        titleProps={{ 'data-testid': 'title' }}
      />,
    );
    expect(getByText('Bannerbase title test')).toHaveClass(
      'mm-banner-base__title',
    );
    expect(getByTestId('title')).toBeDefined();
  });
```

dhruvv173 commented:
hi @georgewrmarshall, what do think about something like this?
```
  it('should render bannerbase children with props', () => {
    const { getByText, getByTestId } = render(
      <BannerBase
        childrenWrapperProps={{
          variant: TextVariant.bodyMd,
          'data-testid': 'childrenWrapper',
        }}
      >
        Bannerbase children test
      </BannerBase>,
    );
    expect(getByTestId('childrenWrapper')).toBeDefined();
    expect(getByText('Bannerbase children test')).toBeDefined();
  });
```

georgewrmarshall commented:
> hi @georgewrmarshall, what do think about something like this?
> 
> ```
>   it('should render bannerbase children with props', () => {
>     const { getByText, getByTestId } = render(
>       <BannerBase
>         childrenWrapperProps={{
>           variant: TextVariant.bodyMd,
>           'data-testid': 'childrenWrapper',
>         }}
>       >
>         Bannerbase children test
>       </BannerBase>,
>     );
>     expect(getByTestId('childrenWrapper')).toBeDefined();
>     expect(getByText('Bannerbase children test')).toBeDefined();
>   });
> ```

Looks good!

dhruvv173 commented:
> > > Hey @dhruvv173, this PR should fix the linting error #20448 and looks like I missed some snapshots or they were updated in the last merge from develop. I was also thinking that because I added the `childrenWrapperProps` we should write a unit test for that. There should be some examples of this with `titleProps` in the unit tests. Would you mind updating it?
> > 
> > 
> > I have updated the snapshot locally and lint error will be fixed once your PR is merged. About the latter part, I just want to confirm that we want to add a unit test which checks for `childrenWrapperProps` in the component, right? If yes, I will check for examples in unit tests and work on it. Thank you!
> 
> Yes, that's correct! I noticed that we aren't actually testing for `titleProps` or `descriptionProps` we can improve the coverage by adjusting each unit test for these props slightly
> 
> ```
> it('should render bannerbase title', () => {
>   const { getByText, getByTestId } = render(
>     <BannerBase
>       title="Bannerbase title test"
>       titleProps={{ 'data-testid': 'title' }}
>     />,
>   );
>  
>   expect(getByText('Bannerbase title test')).toHaveClass('mm-banner-base__title');
>   expect(getByTestId('title')).toBeDefined();
> });;
> ```
> 
> Changing this element to a `tag: 'p'` instead of `h5` tag should also help with the e2e tests
> 
> https://github.com/MetaMask/metamask-extension/blob/b0fcb12b99dd82316c7d26d073efa653bf4b0f80/test/e2e/tests/security-provider.spec.js#L109

every instance of `h5` should be replaced with `p`, right?

dhruvv173 committed:
updating snapshot of blockaid-banner-alert and adding unit test for childrenWrapperProps

georgewrmarshall committed:
BannerBase updates to stories, adding locale for close button, removing static data-testid in favor of using it at the instance, updating snapshots associated with those changes

georgewrmarshall commented on test/e2e/swaps/shared.js:
```diff
@@ -157,7 +157,7 @@ const checkActivityTransaction = async (driver, options) => {
 const checkNotification = async (driver, options) => {
   const boxTitle = await driver.findElement(
-    '[data-testid="mm-banner-base-title"]',
+    '[data-testid="swap-token-verification-banner-title"]',
```
Removing the static data-testid that was added to the component in favor of using `descriptionProps` and making the test id more specific"

georgewrmarshall commented on ui/components/component-library/banner-base/banner-base.tsx:
```diff
@@ -0,0 +1,94 @@
+          {children && typeof children === 'object' ? (
+            children
+          ) : (
+            <Text {...childrenWrapperProps}>{children}</Text>
```
adds `childrenWrapperProps` here

georgewrmarshall commented on ui/components/component-library/banner-base/banner-base.tsx:
```diff
@@ -0,0 +1,94 @@
+
+        <div>
+          {title && (
+            <Text variant={TextVariant.bodyLgMedium} {...titleProps}>
```
Removes static data-testid from title to enable specificity of test id during implementation

georgewrmarshall commented on ui/pages/swaps/prepare-swap-page/prepare-swap-page.js:
```diff
@@ -1036,6 +1036,9 @@ export default function PrepareSwapPage({
                   ? t('swapTokenVerifiedOn1SourceTitle')
                   : t('swapTokenAddedManuallyTitle')
               }
+              titleProps={{
```
Here we are removing the static data-testid from the `BannerBase` component and utilizing `titleProps` to assign one at implementation

georgewrmarshall committed:
Merge branch 'develop' into fix/19181

georgewrmarshall commented on ui/components/component-library/banner-base/banner-base.stories.tsx:
```diff
@@ -39,20 +28,29 @@ export default {
+    childrenProps: {
+      control: 'object',
+    },
     action: {
```
This shouldn't be here remove

georgewrmarshall committed:
Removing incorrect arg from storybook file

georgewrmarshall reviewed (dismissed):
LGTM! I made some small updates to tests and updated the component to remove the static test id and a class name that I don't think we needed. Just need tests to pass. Thanks for your contribution @dhruvv173

garrettbear reviewed (commented):
Nice! LGTM. Just looks like we need to sort out some e2e tests that are failing

georgewrmarshall committed:
Updating html element in security provider e2e test to match BannerBase. Also updating snapshot of ConfirmTransaction page

georgewrmarshall committed:
Fixing e2e tests
